app/Models/DailyChallengeUserStats.php

```diff
@@ -47,27 +48,20 @@ public static function calculate(CarbonImmutable $date): void
             return;
         }
 
-        $highScores = $playlist
+        $highScoresByUserId = $playlist
             ->highScores()
             ->where('total_score', '>', 0)
             ->orderBy('total_score', 'DESC')
-            ->get();
-        $count = $highScores->count();
-        // these variables are only used if there's anything in the array
-        if ($count > 0) {
-            $top50p = $highScores[max(0, (int) ($count * 0.5) - 1)]->total_score;
-            $top10p = $highScores[max(0, (int) ($count * 0.1) - 1)]->total_score;
-        }
-        $highScoresByUserId = [];
-        foreach ($highScores as $highScore) {
-            $highScoresByUserId[$highScore->user_id] = $highScore;
-        }
+            ->get()
+            ->keyBy('user_id');
         $statsByUserId = static
             ::where('last_weekly_streak', '>=', $previousWeek->subDays(1))
-            ->orWhereIn('user_id', array_keys($highScoresByUserId))
+            ->orWhereIn('user_id', $highScoresByUserId->keys())
             ->get()
             ->keyBy('user_id');
-        $userIds = new Set([...$statsByUserId->keys(), ...array_keys($highScoresByUserId)]);
+        $percentile = $playlist->scorePercentile();
```

Comment:
this means it's doing the `$playlist->highScores()->where('total_score', '>', 0)->orderBy` query twice, right?

Reply:
yes. it's a once a day thing and indexed columns and the other only selects one column, so 🤷